src/realm/group_writer.cpp

```diff
@@ -301,25 +302,283 @@ GroupWriter::MapWindow* GroupWriter::get_window(ref_type start_ref, size_t size)
+    };
+
+
+    auto backdate_single_entry = [&](FreeSpaceEntry& entry, bool referenced) -> void {
+        static_cast<void>(referenced);
+        while (entry.released_at_version) {
+            // early out for references before oldest freelist:
+            if (entry.released_at_version <= this->m_oldest_reachable_version) {
+#ifdef REALM_DEBUG
+                REALM_ASSERT(!referenced);
+#endif
+                break;
+            }
+            auto earlier_it = get_earlier(entry.released_at_version);
+            // if no earlier versions than that exists, we're done
+            if (earlier_it == nullptr) {
+#if REALM_ALLOC_DEBUG
+                std::cout << "  backdating [" << entry.ref << ", " << entry.size << "] to zero (no earlier freelist)";
+#endif
+#ifdef REALM_DEBUG
+                REALM_ASSERT(!referenced);
+#endif
+                entry.released_at_version = 0;
+                break;
+            }
+            // earlier versions exist, we need to examine the freelist for it
+#if REALM_ALLOC_DEBUG
+            std::cout << " - earlier freelist: " << earlier_it->version;
+#endif
+            auto covering_blocks_released_at = find_cover_for(entry, *earlier_it);
+            if (covering_blocks_released_at == entry.released_at_version) {
+#ifdef REALM_DEBUG
+                REALM_ASSERT(referenced);
+#endif
+                break;
+            }
+#ifdef REALM_DEBUG
+            REALM_ASSERT(!referenced);
+#endif
+            entry.released_at_version = covering_blocks_released_at;
+        }
+    };
+
+
+#ifdef REALM_DEBUG
+    auto reachables = map_reachable();
+#if REALM_ALLOC_DEBUG
+    std::cout << "  Reachable: ";
+    ref_type last = 0;
+    for (auto& i : reachables) {
+        if (i.ref != last) {
+            last = i.ref;
+            std::cout << std::endl << "    " << i.ref << " - " << i.ref + i.size << " : " << i.released_at_version;
+        }
+        else {
+            std::cout << ", " << i.released_at_version;
+        }
+    }
+    std::cout << std::endl << "  Backdating:";
+#endif
+#endif
+    size_t limit = m_not_free_in_file.size();
+    for (size_t index = 0; index < limit; ++index) {
+        auto& entry = m_not_free_in_file[index];
+        bool referenced = false;
+#ifdef REALM_DEBUG
+#if REALM_ALLOC_DEBUG
+        std::cout << std::endl
+                  << "    Considering [" << entry.ref << ", " << entry.size << "]-" << entry.released_at_version
+                  << " {";
+#endif
+        unsigned k;
+        for (k = 0; k < reachables.size(); ++k) {
+            if (reachables[k].ref + reachables[k].size > entry.ref)
+                break;
+        }
+        while (k < reachables.size() && reachables[k].ref + reachables[k].size > entry.ref &&
+               reachables[k].ref < entry.ref + entry.size) {
+#if REALM_ALLOC_DEBUG
+            std::cout << reachables[k].released_at_version << " ";
+#endif
+            ++k;
+            referenced = true;
+        }
```

Comment:
I was about to make two comments, but then I realized that they don't actually apply due to another problem:

-  The `for` loop should be a `upper_bound` so that we don't go N^2 even in debug builds.
-  `reachables[k].ref + reachables[k].size > entry.ref` is guaranteed to be true in the while condition.

Neither applies because `reachables` is sorted by `{ref, version}` not `{end, ...}` and it can have duplicates. But now there is a different problem: because `size`/`end` isn't in the sort, the `while` loop won't actually see all versions that are reachable. Consider the following scenario (using start and end rather than sizes since that makes things simpler):

```
entry = {7, 10};
reachables = {
  {0, 8, 100}, # for loop stops here 
  {1, 4, 200}, # while loop exits here
  {2, 8, 300}, # while loop _should_ see this, but doesn't
}
```

You could sort by `{end, version}` (although I don't think the `version` sorting actually matters). That would at least make the two bullet points above be improvements. While it would handle the scenario that I described above, it would still miss some versions in a different scenario. I'm not sure if there is a solution other than iterating over the entire reachables (so it doesn't need any sorting) and just printing the version of every block that intersects with entry. But that seems outrageously expensive, even for debug builds (since we do this for every entry in the freelist).

Given that you probably only care about whether something is reachable at all, and not how many times it is reachable, how about just turning the `while` into an `if`? Then we are guaranteed that it is only O(freelist * log (reachables)) which is much nicer than O(freelist * reachables).

(Or just remove this printing)

Reply:
Fixed.